### PR TITLE
List display

### PR DESCRIPTION
**Pull request by sbambur**
## Completed displaying book list

- scroll restoration
- removed async logic (now it is just an array)
- lazy-loading
- routing
- time counter (DD:HH:MM)
- main page style

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,3 +1,15 @@
...
+import Book from './components/Book';
+import Library from './components/Library';
+
 export default function App() {
-  return <p>React</p>;
+  return (
+    <Container className='mb-4'>
+      <Routes>
+        <Route path='/' element={<Library />} />
+        <Route path='/book:id' element={<Book />} />
```
Comment (dariausatova) [inline src/App.tsx:11]: use a router with a separate spelling of the identifier from the word router
'/book/:id'

File: src/components/Book.tsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { Button } from 'react-bootstrap';
+import { NavLink, useParams } from 'react-router-dom';
+import useStore from '../hooks/useStore';
+
+export default function Book() {
```
Comment (dariausatova) [inline src/components/Book.tsx:5]: es6

File: src/components/Book.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  const { id } = useParams();
+  const { books } = useStore();
+
+  return (
+    <div className='pt-4'>
+      <h3>{books[Number(id)].title}</h3>
```
Comment (dariausatova) [inline src/components/Book.tsx:11]: you convert the id twice, wouldn't it be better to do it once?

Comment (dariausatova) [inline src/components/Book.tsx:11]: use destructurization

File: src/components/BookItem.tsx [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { memo } from 'react';
+import { IBook } from '../store';
+import { timeCounter } from './utils/timeCounter';
+
+function BookItem({ data }: { data: IBook }) {
```
Comment (dariausatova) [inline src/components/BookItem.tsx:5]: es6 function

File: src/components/BookItem.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  return (
+    <>
```
Comment (dariausatova) [inline src/components/BookItem.tsx:7]: why you need empty tag?

File: src/components/BookItem.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      <div className={`bookCard pt-4 ${data.returnDate ? 'locked' : ''}`}>
```
Comment (dariausatova) [inline src/components/BookItem.tsx:8]: use && instead of ternary operation

File: src/components/BookItem.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        {data.returnDate && (
+          <div className='returnDate_message'>
```
Comment (dariausatova) [inline src/components/BookItem.tsx:10]: use kebab-case for className

File: src/components/BookItem.tsx (hunk 6 above)
Comment (dariausatova) [inline src/components/BookItem.tsx:8]: use destructurization

File: src/components/Library.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
+import { observer } from 'mobx-react-lite';
+import { ChangeEvent, useCallback, useEffect, useState } from 'react';
+import { Col, Row } from 'react-bootstrap';
+import { NavLink } from 'react-router-dom';
+import useStore from '../hooks/useStore';
+import BookItem from './BookItem';
+import { debounce } from './utils/debounce';
+
+function Library() {
```
Comment (dariausatova) [inline src/components/Library.tsx:9]: es6 function

File: src/components/Library.tsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  const { books, bookCounter } = useStore();
+
+  const [searchParam, setSearchParam] = useState('');
+  const [scrollPosition] = useState(() => {
```
Comment (dariausatova) [inline src/components/Library.tsx:13]: const [scrollPosition] = useState(sessionStorage.getItem('scroll-position') || null);
sessionStorage.removeItem('scroll-position');
is it possible to write in this format?

File: src/components/Library.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    const persistedPosition = sessionStorage.getItem('scroll-position');
+    sessionStorage.removeItem('scroll-position');
+    return persistedPosition ? persistedPosition : null;
+  });
+  const [amountBooks, setAmountBooks] = useState(() => {
```
Comment (dariausatova) [inline src/components/Library.tsx:18]: higher example

File: src/components/Library.tsx [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    const persistedAmount = sessionStorage.getItem('books-amount');
+    sessionStorage.removeItem('books-amount');
+    return persistedAmount ? Number(persistedAmount) : 9;
+  });
+  const [currentBooksList, setCurrentBookList] = useState(books.slice(0, amountBooks));
+
+  console.log('render');
```
Comment (dariausatova) [inline src/components/Library.tsx:25]: remove console

File: src/components/Library.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+  const handleInput = (e: ChangeEvent<HTMLInputElement>) => {
```
Comment (dariausatova) [inline src/components/Library.tsx:27]: why do you need to send the whole event here, and not just send the value?

Comment (dariausatova) [inline src/components/Library.tsx:27]: why do you need this function if there is only one action in it?

File: src/components/Library.tsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    setSearchParam(e.target.value);
+  };
+
+  const scrollHandler = useCallback(
+    (e: any) => {
```
Comment (dariausatova) [inline src/components/Library.tsx:32]: don't use any type

File: src/components/Library.tsx [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        if (amountBooks < books.length) {
+          setAmountBooks((prevState) => prevState + 9);
+        }
+      }
+    },
+    [amountBooks, books.length]
+  );
+
+  const setScrolPosition = (scroll: number) => {
+    sessionStorage.setItem('scroll-position', JSON.stringify(scroll));
```
Comment (dariausatova) [inline src/components/Library.tsx:47]: why do you need stringify?

File: src/components/Library.tsx [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+  };
+
+  const debounceSetter = debounce(() => setScrolPosition(window.pageYOffset), 500);
```
Comment (dariausatova) [inline src/components/Library.tsx:50]: setScrollPosition

File: src/components/Library.tsx [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+  useEffect(() => {
+    window.scrollTo(0, Number(scrollPosition));
+  }, []);
+
+  useEffect(() => {
+    sessionStorage.setItem('books-amount', JSON.stringify(amountBooks));
```
Comment (dariausatova) [inline src/components/Library.tsx:57]: why do you need stringify?

File: src/components/Library.tsx [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+    if (books.length !== currentBooksList.length) {
+      setCurrentBookList(books.slice(0, amountBooks));
+    }
+  }, [books, amountBooks, currentBooksList.length]);
+
+  useEffect(() => {
+    document.addEventListener('scroll', debounceSetter);
```
Comment (dariausatova) [inline src/components/Library.tsx:65]: why not write these functions in the effect above?

File: src/components/Library.tsx [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+          className='header_input'
+          type='text'
+          value={searchParam}
+          onChange={handleInput}
+        />
+      </div>
+      <Row md={2} xs={1} lg={3} className='g-3'>
+        {currentBooksList.map((book, index) => (
+          <Col key={book.id}>
+            {book.returnDate ? (
```
Comment (dariausatova) [inline src/components/Library.tsx:87]: I don't see where this [task](https://trello.com/c/1NtoM3Mw/1-%D0%B8%D0%BD%D1%84%D0%BE) is implemented in this place

File: src/components/utils/debounce.ts [hunk 19]
```diff
@@ -0,0 +1,7 @@
+export const debounce = (fn: Function, ms = 300) => {
+  let timeout: ReturnType<typeof setTimeout>;
+  return function (this: any, ...args: any[]) {
```
Comment (dariausatova) [inline src/components/utils/debounce.ts:3]: es6 function

File: src/components/utils/timeCounter.ts [hunk 20]
```diff
@@ -0,0 +1,14 @@
+export const timeCounter = (returnTime: string) => {
```
Comment (dariausatova) [inline src/components/utils/timeCounter.ts:1]: why don't you use moment js?

File: src/context/index.tsx [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+import { BookStore } from '../store';
+
+interface StoreContextProviderProps {
+  children?: ReactNode;
+}
+
+const store = new BookStore();
+export const StoreContext = createContext(store);
+
+export function StoreContextProvider({ children }: StoreContextProviderProps) {
```
Comment (dariausatova) [inline src/context/index.tsx:11]: es6 function

File: src/hooks/useStore.ts [hunk 22]
```diff
@@ -0,0 +1,6 @@
+import { useContext } from 'react';
+import { StoreContext } from '../context';
+
+export default function useStore() {
```
Comment (dariausatova) [inline src/hooks/useStore.ts:4]: es6

File: src/styles.css [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+
+#root {
+  background: linear-gradient(179.27deg, #2DB3DD 0.82%, #FFFFFF 86.37%);
+  background-attachment: fixed;
+  height: 100%;
+  min-height: 100vh;
+  font-family: 'Inter';
+}
+
+.main a{
```
Comment (dariausatova) [inline src/styles.css:11]: add space

File: src/styles.css [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+.returnDate_message > p {
+  margin: 0;
+  font-weight: 700;
+  font-size: 36px;
+  line-height: 44px;
+}
+
+.bookCard_image img{
```
Comment (dariausatova) [inline src/styles.css:90]: add space

File: src/styles.css [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+  width: 180px;
+  height: 180px;
+  object-fit: cover
```
Comment (dariausatova) [inline src/styles.css:93]: ;

File: src/styles.css [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+
+.bookCard_author {
+  font-size: 14px;
+  line-height: 17px;
+  margin: 20px 0;
+}
+
+.bookCard_title {
+  font-size: 32px;
+line-height: 39px;
```
Comment (dariausatova) [inline src/styles.css:104]: add tab

File: src/styles.css [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+}
+
```
Comment (dariausatova) [inline src/styles.css:106]: remove empty lines

File: src/styles.css [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+
+
+.Loading {
```
Comment (dariausatova) [inline src/styles.css:109]: why is the class style different?

File: src/components/utils/timeCounter.ts (hunk 20 above)
Comment (sbambur) [reply src/components/utils/timeCounter.ts:1]: Only one function is needed here, there is no point in the whole library

File: src/components/BookItem.tsx [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+import { NavLink } from 'react-router-dom';
+import { IBook } from '../store';
+import { timeCounter } from './utils/timeCounter';
+
+const BookItem = ({ data }: { data: IBook }) => {
+  const { returnDate, title, author, image, id } = data;
+
+  return (
+    <Col>
+      <NavLink to={`/book/${id}`} className={returnDate ? 'locked' : ''}>
```
Comment (dariausatova) [inline src/components/BookItem.tsx:12]: use && in className

File: src/components/Library.tsx [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+    Number(sessionStorage.getItem('books-amount')) || 9
+  );
+  const [currentBooksList, setCurrentBookList] = useState(books.slice(0, amountBooks));
+
+  sessionStorage.removeItem('books-amount');
+  sessionStorage.removeItem('scroll-position');
+
+  const scrollHandler = useCallback(() => {
+    if (
+      document.documentElement.scrollHeight -
```
Comment (dariausatova) [inline src/components/Library.tsx:20]: put this expression into a variable and combine 2 if into one condition

File: src/styles.css [hunk 31]
```diff
@@ -0,0 +1,114 @@
+@import url('https://fonts.googleapis.com/css2?family=Inter:wght@100;300;400;700&display=swap');
+html {
+  scroll-behavior: unset !important;
+}
+
```
Comment (dariausatova) [inline src/styles.css:5]: remove empty lines

**Pull request merged**